modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2ScopeGrantLocalServiceImpl.java

```diff
@@ -118,18 +118,22 @@ public Collection<OAuth2ScopeGrant> getOAuth2ScopeGrants(
 		long companyId, String applicationName, String bundleSymbolicName,
 		String accessTokenContent) {
 
-		OAuth2Authorization oAuth2Authorization =
-			_oAuth2AuthorizationLocalService.
-				fetchOAuth2AuthorizationByAccessTokenContent(
-					accessTokenContent);
+		List<OAuth2Authorization> oAuth2Authorizations =
+			oAuth2AuthorizationPersistence.findByC_ATCH(
+				companyId, accessTokenContent.hashCode());
 
-		if (oAuth2Authorization == null) {
+		if (oAuth2Authorizations.isEmpty()) {
 			return Collections.emptyList();
 		}
 
-		List<OAuth2ScopeGrant> oAuth2ScopeGrants =
-			oAuth2ScopeGrantPersistence.getOAuth2AuthorizationOAuth2ScopeGrants(
-				oAuth2Authorization.getOAuth2AuthorizationId());
+		List<OAuth2ScopeGrant> oAuth2ScopeGrants = new ArrayList<>();
+
+		for (OAuth2Authorization oAuth2Authorization : oAuth2Authorizations) {
+			oAuth2ScopeGrants.addAll(
+				oAuth2ScopeGrantPersistence.
+					getOAuth2AuthorizationOAuth2ScopeGrants(
+						oAuth2Authorization.getOAuth2AuthorizationId()));
+		}
 
 		oAuth2ScopeGrants.removeIf(
```

Comment:
This is bad for performance, you should not collect everything then purge the non-matching entries. The logic should be flipped to only collect matching entries for returning.

Reply:
your method should be like this:
```
	@Override
	public Collection<OAuth2ScopeGrant> getOAuth2ScopeGrants(
		long companyId, String applicationName, String bundleSymbolicName,
		String accessTokenContent) {

		List<OAuth2ScopeGrant> oAuth2ScopeGrants = new ArrayList<>();

		for (OAuth2Authorization oAuth2Authorization :
				oAuth2AuthorizationPersistence.findByC_ATCH(
					companyId, accessTokenContent.hashCode())) {

			if (!Objects.equals(
					accessTokenContent,
					oAuth2Authorization.getAccessTokenContent())) {

				continue;
			}

			for (OAuth2ScopeGrant oAuth2ScopeGrant :
					oAuth2ScopeGrantPersistence.
						getOAuth2AuthorizationOAuth2ScopeGrants(
							oAuth2Authorization.getPrimaryKey())) {

				if (Objects.equals(
						applicationName,
						oAuth2ScopeGrant.getApplicationName()) &&
					Objects.equals(
						bundleSymbolicName,
						oAuth2ScopeGrant.getBundleSymbolicName())) {

					oAuth2ScopeGrants.add(oAuth2ScopeGrant);
				}
			}
		}

		return oAuth2ScopeGrants;
	}
```